compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -1955,6 +1964,410 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+    /// - Find all of the generics that are mentioned in the predicate (`Filling`).
+    /// - Examine the `Self` type in the `impl`, and see which of its type argument(s) mention any of those generics.
+    /// - Examing the definition for the `Self` type, and identify (for each of its variants) if there's a unique field
+    ///   which uses those generic arguments.
+    /// - If there is a unique field mentioning the "blameable" arguments, use that field for the error span.
+    ///
+    /// Before we do any of this logic, we recursively call `point_at_specific_expr_if_possible` on the parent
+    /// obligation. Hence we refine the `expr` "outwards-in" and bail at the first kind of expression/impl we don't recognize.
+    ///
+    /// This function returns a `Result<&Expr, &Expr>` - either way, it returns the `Expr` whose span should be
+    /// reported as an error. If it is `Ok`, then it means it refined successfull. If it is `Err`, then it may be
+    /// only a partial success - but it cannot be refined even further.
+    ///
+    /// `point_at_specific_expr_if_possible` will ultimately return the returned `Expr` whether it is `Ok` or `Err`;
+    /// this error state is only used for propagation within these recursive functions.
+    fn point_at_specific_expr_if_possible_for_derived_predicate_obligation(
+        &self,
+        obligation: &traits::ImplDerivedObligationCause<'tcx>,
+        expr: &'tcx hir::Expr<'tcx>,
+    ) -> Result<&'tcx hir::Expr<'tcx>, &'tcx hir::Expr<'tcx>> {
+        // First, we attempt to refine the `expr` for our span using the parent obligation.
+        // If this cannot be done, then we are already stuck, so we stop early (hence the use
+        // of the `?` try operator here).
+        let expr = self.point_at_specific_expr_if_possible_for_obligation_cause_code(
+            &*obligation.derived.parent_code,
+            expr,
+        )?;
+
+        // These are the generic parameters used by the `impl` that produced our current obligation.
+        let impl_generics: &ty::Generics = self.tcx.generics_of(obligation.impl_def_id);
+
+        // This is the "trait" (meaning, the predicate "proved" by this `impl`) which provides the `Self` type we care about.
+        // For the purposes of this function, we hope that it is a `struct` type, and that our current `expr` is a literal of
+        // that struct type.
+        let impl_trait_self_ref: Option<ty::TraitRef<'tcx>> =
+            self.tcx.impl_trait_ref(obligation.impl_def_id);
+
+        let Some(impl_trait_self_ref) = impl_trait_self_ref else {
+            // It is possible that this is absent. In this case, we make no progress.
+            return Err(expr);
+        };
+
+        // We only really care about the `Self` type itself, which we extract from the ref.
+        let impl_self_ty: Ty<'tcx> = impl_trait_self_ref.self_ty();
```

Comment:
This is a totally a tiny nitpick, but unless these type ascriptions are guiding inference, they're kinda an eyesore. Do you mind removing them?

```suggestion
        let impl_self_ty = impl_trait_self_ref.self_ty();
```

Reply:
Ah, I have some issue where `rust-analyzer` doesn't pick up all of the types in the rustc codebase (it seems to fail on some of the query functions?) So it can compile without them, but VSCode stops giving type-directed hints without them, which makes it harder for me to figure out what I'm doing.

I can remove them in a later pass; I think it's going to take me a few runs through this PR to remove all of the clunkiness.

Reply:
No worries -- thanks, and yeah, this can be dealt with in the end.

Reply:
There were a few problems with types, but if you use this config and rebase to the latest master it should all work https://rustc-dev-guide.rust-lang.org/building/suggested.html#configuring-rust-analyzer-for-rustc